add option to view full (html) message

r10s opened the pull request:
this is the ui counterpart of https://github.com/deltachat/deltachat-core-rust/pull/2125 - the pr adds a button "Show full message..." whenever core thinks, this is reasonable. a click on the button opens an html-view with the full message as returned by the core.

the full-message is even searchable :)

- [x] adapt to api
- [x] add button if core says so
- [x] show html-mails in webview
- [x] make mailto: links work
- [x] add option to downoad remote content, defaults of "never" for now

<img width=320 src=https://user-images.githubusercontent.com/9800740/103411283-6e5d6680-4b6f-11eb-8200-50afa5631cdc.png> <img width=320 src=https://user-images.githubusercontent.com/9800740/103411281-6bfb0c80-4b6f-11eb-9ed5-11d179a05cb3.png>
_left: message with the new button, right: after tap on the button_

EDIT: now, remote content is disabled by default and can be enabled in the menu:

<img width=320 src=https://user-images.githubusercontent.com/9800740/105777407-9a9bb580-5f6a-11eb-87f2-3c8c0ab50ab4.png>

related related: https://github.com/deltachat/deltachat-desktop/issues/2078, https://github.com/deltachat/deltachat-ios/issues/1046, https://github.com/deltachat/deltachat-core-rust/pull/2125

Hocuri commented:
- When fetching from remote servers, we have to think a lot about security, e.g. [efail](https://efail.de/) ([on Heise](https://www.heise.de/security/meldung/PGP-E-Mail-Verschluesselung-akut-angreifbar-4048489.html)) and I must say, it was a nice feeling to know that there is no way DC could be affected by such security problems
- Is javascript enabled? (also, do newsletters even need javascript)
- We now have 3 possible ways of showing a message:
  - plain-text, as we currently do it (of course we will still do this, but if start cropping more, the user might be forced to hit "show full message" even though there are now images at all)
  - html but fetching from a remote server disabled
  - html, fetching images from a remote server enabled
  
  It probably won't be possible UX-wise to always let the user choose between those, but what about this:
  - By default, disable all remote-fetching of images in full-message-view. But add a button to enable it. Then, directly enable it for all messages from this contact, but add a possibility to hide images again. Thunderbird writes "To protect your privacy, Thunderbird has blocked remote content in this message." and just adds a "Preferences" drop-down menu. K9 directly shows a "SHOW PICTURES" button.

Simon-Laux commented:
> Is javascript enabled? (also, do newsletters even need javascript)

hopefully not, js is not available in email clients for a good reason.


> html but fetching from a remote server disabled

remote fetching should be controllable by the user, like a load external content button and the option to trust certain domains / filetypes?

Also we should sanitize the html files even if we disable js in the webview there are probably still vulnerabilities. I would just allow a safe subset of html.

r10s commented:
sanitizing html-files is planned, filtering out js stuff and maybe other things, this will be done together with adding embedded images which requires iterating over html anyway.

for loading remote content - we had several discussions over that, there are several possible approaches, yes.

r10s commented:
> Is javascript enabled? (also, do newsletters even need javascript)

no, it is not. for subsequent discussions wrt sanitizing, please refer to https://github.com/deltachat/deltachat-core-rust/issues/2127

r10s commented:
remote content is no disabled by default, there is a menu option to load it once or always, see image atop.

if needed, we can add a banner or so later on in a subsequent pr, however, for the first thing, the state is not that bad, esp. as there are not many options currently :)

Hocuri commented:
First testing...

- How hard would it be to allow zooming?
- Currently the user has to hit "Load remote content" to show images. This will prevent users to accidentally give up their privacy and load images, but users might also not find this button. (just noted, I don't have a strong opinion on this)
- +1 for blocking loading images via http (I first wondered why some images are not shown but images that come via http don't deserve to be shown)
- The "Always" option means "Always for this sender", doesn't it? 
- it took me a while to understand that the check means "current selection"
  - For the last two points, maybe make radio buttons so that 1. we can write "Always for this sender" 2. it's clear how to communicate the currently-selected option? disadvantage: We need a custom view (OTOH, we have experience with custom views, so adding them will be faster :) )

r10s commented:
> How hard would it be to allow zooming?

probably not that hard, i've skimmed some options `WebSettings`, but this is for another pr then.

> users might also not find this button.

we will see. it is one of two menu entries.

i also noted above that a banner in a subsequent pr might be nice, however, this would also need detection of request to remote content, not sure how hard that is and if it is really worth the effort.

> The "Always" option means "Always for this sender", doesn't it?

now, "Always" means "Always" :) so, "Always for all HTML-messages".

at least for now, that should be sufficient. tbh, i do not think we really need this "always for this message", "always for this sender", "always for this domain" etc., i do not thing that this bureaucracy is really helpful. also, as you usually have confirmed a contact request before and you _always_ have to hit the "show full message" button. having that in mind, that the html-stuff is used for confirmed contacts only, the privacy thing is maybe not a huge one.

but i would like to get this discussion out of this pr as well, the current options are already more than we initially wanted to do :)

Hocuri commented on res/layout/conversation_item_received.xml:
```diff
@@ -151,6 +151,23 @@
+                android:layout_marginBottom="10dp"
+                android:paddingLeft="12dp"
+                android:paddingRight="12dp"
+                android:background="@drawable/conversation_item_update_background"
```
The button somehow looks too... grey to me; the following looks better IMO but not great still:

```suggestion
                android:background="@color/delta_primary"
                android:textColor="@color/white"
                tools:visibility="visible"
```

r10s replied:
yip, the button could need some tweaking, however, also outgoing buttons need tweaking, and by just saying `@color/delta_primary`, the rounded corners are lost.

Hocuri replied:
For me it looks equally good with round and square corners, but yes, if we want to preserve rounded corners, this needs some more work.

What do you mean with "outgoing buttons"?

r10s replied:
> What do you mean with "outgoing buttons"? 

the ones defined in `res/layout/conversation_item_sent.xml`. we have two layout for the bubbles, one incoming, one outgoing. you suggestion would only change the incoming ones (on the left side).

however, to move forward, i'd like to put that layout discussion into another pr, current state is not bad.

r10s commented:
i've changed the wording from "Load remote content" to "Load remote images" - for the latter, it seems much easier for the user to understand the option as one sees that images are missing (as in the screenshot below).

not sure, if it is really needed to point out that this also affects styles or other content, however, i've added that bit. 

the second note is that user are not thinking that it is a bug, when they are seeing images if the option is disabled (i've seen that complain for other mua in some threads and forums)

<img width=320 src=https://user-images.githubusercontent.com/9800740/106030270-4c9fc280-60ce-11eb-986a-1a407dadf68f.png>

link2xt commented:
Shorter the messages are easier to read and some users have small screens. Maybe "Remote images may be used to track you" for the first sentence?

Typical WYSIWYG users probably will understand what "remote styles" are, unless they are familiar with CSS and the concept of separate styles. I would use "fonts *and* other content" instead.

Bullets are useless here, this is not a list. I would even merge it into a single paragraph: "The setting also allows loading fonts and other content. Even if it is disabled, you may still see embedded or cached images."

r10s commented:
now it looks like that, much cleaner, thanks a lot!

<img width=320 src=https://user-images.githubusercontent.com/9800740/106044404-fa679d00-60df-11eb-9f2d-d671d1410faf.png>

Hocuri commented on src/org/thoughtcrime/securesms/LocalHelpActivity.java:
```diff
@@ -1,269 +1,80 @@
+      exists = true;
+      is.close();
+    } catch(Exception e) {
+      ;
```
Maybe forgot to print the exception? (if it happens from time to time and you don't want to spam the log, maybe also use `Log.w(TAG, e);`, this will only print just a single line)

r10s replied:
errors thrown in this function are normal an expected, we are using it to check for existance, as the "normal" ways to check for assets are erroneous. logging does not make much sense therefore, you would get warnings/info everytime you open the help. this code is also not changed by this pr, it was just copied here (sorry, i should have mention that: LocalHelpActivity was included in WebViewActivity before.

r10s replied:
but adding a comment makes sense, i do a commit.

r10s commented on src/org/thoughtcrime/securesms/FullMsgActivity.java:
```diff
@@ -0,0 +1,147 @@
+        // not-always-mode where the dialog is used more often. Or [Ok] would mean "Once" as well as "Change checkbox setting",
+        // which is also a bit weird. Anyway, let's give the three buttons a try :)
+        if (Prefs.getAlwaysLoadRemoteContent(this)) {
+          builder.setNeutralButton("✔️ " + this.getString(R.string.always), (dialog, which) -> onChangeLoadRemoteContent(LoadRemoteContent.ALWAYS));
+          builder.setNegativeButton(R.string.never, (dialog, which) -> onChangeLoadRemoteContent(LoadRemoteContent.NEVER));
+        } else {
+          builder.setNeutralButton(R.string.always, (dialog, which) -> onChangeLoadRemoteContent(LoadRemoteContent.ALWAYS));
+          builder.setNegativeButton("✔️ " + this.getString(R.string.never), (dialog, which) -> onChangeLoadRemoteContent(LoadRemoteContent.NEVER));
```
have you tested that on your device? can you post a screenshot? on android i am using here, these checkmarks are blue and contrast to dark as well as to light mode. but even black should not be that bad as the background is grey (and they are not so super-important anyway)

Hocuri replied:
![Screenshot_20210128-162052 (1)](https://user-images.githubusercontent.com/18012815/106159339-f772b800-6184-11eb-8d00-9969ae926f38.png)
![Screenshot_20210128-161915 (2)](https://user-images.githubusercontent.com/18012815/106159361-fb9ed580-6184-11eb-8678-7d096e1e0887.png)
(Fairphone 3, Android 10) 
Strange...

r10s replied:
k, thanks. another font or so. seems there are more fonts where that is black, https://emojipedia.org/check-mark/, so going for https://emojipedia.org/check-box-with-check/ in dark mode (as you suggested) seems to be reasonable.

btw. how does that toast when you forward a message to "Saved messages" look on your device? that is using the same checkmark.

r10s replied:
nevermind, i've added `DynamicTheme.getCheckmarkEmoji()` that is also use for that toast.

Hocuri replied:
The toast actually looks ok as toasts are always white-background on my device -> good contrast

Hocuri replied:
But now it looks good!

Hocuri replied:
+1, yeah, probably better to always use https://emojipedia.org/check-box-with-check/ in dark mode, in case the toast background is dark (because this checkmark always has good contrast)

r10s commented on src/org/thoughtcrime/securesms/FullMsgActivity.java:
```diff
@@ -0,0 +1,147 @@
+          builder.setNeutralButton(R.string.always, (dialog, which) -> onChangeLoadRemoteContent(LoadRemoteContent.ALWAYS));
+          builder.setNegativeButton("✔️ " + this.getString(R.string.never), (dialog, which) -> onChangeLoadRemoteContent(LoadRemoteContent.NEVER));
+        }
+        builder.setPositiveButton(R.string.once, (dialog, which) -> onChangeLoadRemoteContent(LoadRemoteContent.ONCE));
```
yip, that makes sense.

r10s commented on src/org/thoughtcrime/securesms/ConversationListActivity.java:
```diff
@@ -282,7 +282,7 @@ public void openConversation(int chatId, int startingPosition) {
     if (isForwarding(this) && dcContext.getChat(chatId).isSelfTalk()) {
       SendRelayedMessageUtil.immediatelyRelay(this, chatId);
-      Toast.makeText(this, "✔️ " + getString(R.string.saved), Toast.LENGTH_SHORT).show();
+      Toast.makeText(this, DynamicTheme.getCheckmarkEmoji(this) + "️ " + getString(R.string.saved), Toast.LENGTH_SHORT).show();
```
this is not related to html-view, however, during review it turns out that the "✔️" emoji does not contrast nicely to dark-mode background-colors and we introduced a function therefore - and used it at places where  "✔️" was already used before.

r10s commented on src/org/thoughtcrime/securesms/ConversationListActivity.java:
```diff
@@ -282,7 +282,7 @@ public void openConversation(int chatId, int startingPosition) {
     if (isForwarding(this) && dcContext.getChat(chatId).isSelfTalk()) {
       SendRelayedMessageUtil.immediatelyRelay(this, chatId);
-      Toast.makeText(this, "✔️ " + getString(R.string.saved), Toast.LENGTH_SHORT).show();
+      Toast.makeText(this, DynamicTheme.getCheckmarkEmoji(this) + " " + getString(R.string.saved), Toast.LENGTH_SHORT).show();
```
this is not related to html-view, however, during review it turns out that the "✔️" emoji does not contrast nicely to dark-mode background-colors and we introduced a function therefore - and used it at places where "✔️" was already used before.

r10s commented on res/menu/web_view.xml:
```diff
@@ -0,0 +1,27 @@
+
+<menu xmlns:android="http://schemas.android.com/apk/res/android" xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <item android:title="@string/search"
+        android:id="@+id/menu_search_web_view"
+        app:actionViewClass="androidx.appcompat.widget.SearchView"
+        app:showAsAction="collapseActionView|never" />
```
we are using the same order already at other places, so i would not mix that.

i read it as "menu_search_web_view" introduces the sequence of needed search buttons :)
